packages/react-hooks/src/useSubscription/useSubscription.ts

```diff
@@ -0,0 +1,55 @@
+import { MedplumClient, SubscriptionEmitter, SubscriptionEventMap } from '@medplum/core';
+import { Bundle } from '@medplum/fhirtypes';
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { useMedplum } from '../MedplumProvider/MedplumProvider.context';
+
+const SUBSCRIPTION_DEBOUNCE_MS = 3000;
+
+export function useSubscription(criteria: string, callback: (bundle: Bundle) => void): void {
+  const medplum = useMedplum() as MedplumClient;
+  const [emitter, setEmitter] = useState<SubscriptionEmitter>();
+
+  const listeningRef = useRef(false);
+  const unsubTimerRef = useRef<ReturnType<typeof setTimeout>>();
+  const prevCriteriaRef = useRef<string>();
+
+  const callbackRef = useRef<typeof callback>();
+  callbackRef.current = callback;
+
+  useEffect(() => {
+    if (unsubTimerRef.current) {
+      clearTimeout(unsubTimerRef.current);
+      unsubTimerRef.current = undefined;
+    }
+    if (prevCriteriaRef.current !== criteria) {
+      setEmitter(medplum.subscribeToCriteria(criteria));
+    }
+    return () => {
+      unsubTimerRef.current = setTimeout(() => {
+        setEmitter(undefined);
+      }, SUBSCRIPTION_DEBOUNCE_MS);
+    };
+  }, [medplum, criteria]);
+
+  useEffect(() => {
+    prevCriteriaRef.current = criteria;
+  }, [criteria]);
+
+  const emitterCallback = useCallback((event: SubscriptionEventMap['message']) => {
```

Comment:
What is the advantage of this over calling `callbackRef.current()` directly?

Reply:
Having to rebind any time the callback changes, otherwise every rerender the event listener has a callback bound and unbound. In this way, we can bind one time and then it doesn't matter even if the callback is a new function every render